pkg/cmd/grafana-cli/commands/datamigrations/migrate_secrets.go

```diff
@@ -0,0 +1,201 @@
+			continue
+		}
+
+		decrypted, err := secretsSrv.DecryptJsonData(context.Background(), row.SecureJsonData)
+		if err != nil {
+			return err
+		}
+
+		var toUpdate struct {
+			SecureJsonData map[string][]byte
+		}
+
+		toUpdate.SecureJsonData, err = secretsSrv.EncryptJsonDataWithDBSession(context.Background(), decrypted, secrets.WithoutScope(), sess)
+		if err != nil {
+			return err
+		}
+
+		if _, err := sess.Table(s.tableName).Update(toUpdate); err != nil {
+			return err
+		}
+	}
+
+	logger.Infof("Secure json data from %s has been re-encrypted successfully\n", s.tableName)
+
+	return nil
+}
+
+type alertingSecret struct{}
+
+func (s alertingSecret) migrate(secretsSrv *manager.SecretsService, sess *sqlstore.DBSession) error {
+	var results []struct {
+		Id                        int
+		AlertmanagerConfiguration []byte
+	}
+
+	selectSQL := "SELECT id, alertmanager_configuration FROM alert_configuration"
+	if err := sess.SQL(selectSQL).Find(&results); err != nil {
+		return err
+	}
+
+	for _, result := range results {
+		postableUserConfig, err := notifier.Load(result.AlertmanagerConfiguration)
+		if err != nil {
+			return err
+		}
+
+		for _, receiver := range postableUserConfig.AlertmanagerConfig.Receivers {
+			for _, gmr := range receiver.GrafanaManagedReceivers {
+				for k, v := range gmr.SecureSettings {
+					decoded, err := base64.StdEncoding.DecodeString(v)
+					if err != nil {
+						return err
+					}
+
+					decrypted, err := secretsSrv.Decrypt(context.Background(), decoded)
+					if err != nil {
+						return err
+					}
+
+					reencrypted, err := secretsSrv.EncryptWithDBSession(context.Background(), decrypted, secrets.WithoutScope(), sess)
+					if err != nil {
+						return err
+					}
+
+					gmr.SecureSettings[k] = base64.StdEncoding.EncodeToString(reencrypted)
+				}
+			}
+		}
+
+		result.AlertmanagerConfiguration, err = json.Marshal(postableUserConfig)
+		if err != nil {
+			return err
+		}
+
+		if _, err := sess.Table("alert_configuration").Update(&result); err != nil {
+			return err
+		}
+	}
+
+	logger.Info("Alerting secrets has been re-encrypted successfully\n")
+
+	return nil
+}
+
+func MigrateSecrets(_ utils.CommandLine, runner runner.Runner) error {
+	if !runner.SettingsProvider.IsFeatureToggleEnabled("envelopeEncryption") {
+		logger.Warn("Envelope encryption is not enabled, quitting...")
+		return nil
+	}
+
+	toMigrate := []interface {
+		migrate(*manager.SecretsService, *sqlstore.DBSession) error
+	}{
+		simpleSecret{tableName: "dashboard_snapshot", columnName: "dashboard_encrypted", isBase64Encoded: false},
+		simpleSecret{tableName: "user_auth", columnName: "o_auth_access_token", isBase64Encoded: true},
+		simpleSecret{tableName: "user_auth", columnName: "o_auth_refresh_token", isBase64Encoded: true},
+		simpleSecret{tableName: "user_auth", columnName: "o_auth_token_type", isBase64Encoded: true},
+		jsonSecret{tableName: "data_source"},
+		jsonSecret{tableName: "plugin_setting"},
+		alertingSecret{},
```

Comment:
Dear reviewer, could you double-check these are all the secrets to migrate and that only the ones specified as (`isBase64Encoded: true`) are the ones that need to be decoded & encoded apart from reencrypted, please?

One way to double-check that could be:
- Looking at `secrets.Service` usages to see which parts of the code use the encryption methods, so all the database columns that are related with these usages should be present on this list.
- Looking at those usages (for example, when getting, creating and/or updating data sources, plugin settings, etc) and checking if that process encodes/decodes the data apart from encrypting/decrypting it.

Of course I did multiple tests, but love to have someone else double check.

Thanks!! ❤️

Reply:
The list seems right to me. 

The only new place I've discovered is this service https://github.com/grafana/grafana/blob/main/pkg/services/live/pipeline/storage_file.go#L20: it uses encryption service instead of secrets service thus far, but probably it should be switched to use envelope encryption. I'm not sure about it though. WDYT @joanlopez @sakjur ?

Reply:
Since this has already been addressed here: https://github.com/grafana/grafana/pull/41625 and looks like it does not need any kind of migration.

Do you think we can resolve this point and go ahead? 🤔 

Thanks!

Reply:
sure! thanks